hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBaseStruct.scala

```diff
@@ -0,0 +1,124 @@
+package is.hail.expr.types.physical
+
+import is.hail.annotations.{CodeOrdering, Region, RegionValue, UnsafeOrdering, UnsafeUtils}
+import is.hail.asm4s.{Code, const, _}
+import is.hail.expr.ir.{EmitMethodBuilder, SortOrder}
+import is.hail.expr.types.BaseStruct
+import is.hail.utils._
+
+object PCanonicalBaseStruct {
+  def getByteSizeAndOffsets(types: IndexedSeq[PType], nMissingBytes: Long, byteOffsets: Array[Long]): Long = {
+    assert(byteOffsets.length == types.length)
+    val bp = new BytePacker()
+
+    var offset: Long = nMissingBytes
+    types.zipWithIndex.foreach { case (t, i) =>
+      val fSize = t.byteSize
+      val fAlignment = t.alignment
+
+      bp.getSpace(fSize, fAlignment) match {
+        case Some(start) =>
+          byteOffsets(i) = start
+        case None =>
+          val mod = offset % fAlignment
+          if (mod != 0) {
+            val shift = fAlignment - mod
+            bp.insertSpace(shift, offset)
+            offset += (fAlignment - mod)
+          }
+          byteOffsets(i) = offset
+          offset += fSize
+      }
+    }
+    offset
+  }
+
+  def alignment(types: Array[PType]): Long = {
+    if (types.isEmpty)
+      1
+    else
+      types.map(_.alignment).max
+  }
+}
+
+trait PCanonicalBaseStruct extends PBaseStruct {
+  lazy val fieldRequired: Array[Boolean] = types.map(_.required)
+
+  lazy val fieldNames: Array[String] = fields.map(_.name).toArray
+
+  val missingIdx = new Array[Int](size)
```

Comment:
Moved missingIdx through alignment here because the implementations on PCanonicalTuple and PCanonicalStruct are identical

Reply:
(and because they were never defined on PBaseStruct)